drivers/i2c/i2c_nrfx_twi_rtio.c

```diff
@@ -0,0 +1,365 @@
+	};
+	uint32_t xfer_flags = 0;
+	nrfx_err_t res;
+
+	/* In case the STOP condition is not supposed to appear after
+	 * the current message, check what is requested further:
+	 */
+	if (!(flags & I2C_MSG_STOP)) {
+		/* - if the transfer consists of more messages
+		 *   finish the transfer without generating the STOP
+		 *   condition, unless the current message is an RX
+		 *   request, for which such feature is not supported
+		 */
+		if (flags & I2C_MSG_READ) {
+			ret = -ENOTSUP;
+		} else {
+			xfer_flags |= NRFX_TWI_FLAG_TX_NO_STOP;
+		}
+	}
+
+	if (!ret) {
+		res = nrfx_twi_xfer(&config->twi, &cur_xfer, xfer_flags);
+		switch (res) {
+		case NRFX_SUCCESS:
+			break;
+		case NRFX_ERROR_BUSY:
+			ret = -EBUSY;
+			break;
+		default:
+			ret = -EIO;
+			break;
+		}
+	}
+
+	if (ret != 0) {
+		nrfx_twi_disable(&config->twi);
+		k_sem_give(&dev_data->completion_sync);
+
+		return i2c_rtio_complete(ctx, ret);
+	}
+
+	return false;
+}
+
+static int i2c_nrfx_twi_configure(const struct device *dev,
+				  uint32_t dev_config)
+{
+	const struct i2c_nrfx_twi_config *config = dev->config;
+	struct i2c_nrfx_twi_data *data = dev->data;
+	nrfx_twi_t const *inst = &config->twi;
+
+	if (I2C_ADDR_10_BITS & dev_config) {
+		return -EINVAL;
+	}
+
+	switch (I2C_SPEED_GET(dev_config)) {
+	case I2C_SPEED_STANDARD:
+		nrf_twi_frequency_set(inst->p_twi, NRF_TWI_FREQ_100K);
+		break;
+	case I2C_SPEED_FAST:
+		nrf_twi_frequency_set(inst->p_twi, NRF_TWI_FREQ_400K);
+		break;
+	default:
+		LOG_ERR("unsupported speed");
+		return -EINVAL;
+	}
+	data->dev_config = dev_config;
+
+	return 0;
+}
+
+static bool i2c_nrfx_twi_start(const struct device *dev)
+{
+	struct i2c_nrfx_twi_data *const dev_data = dev->data;
+	struct i2c_rtio *ctx = dev_data->ctx;
+	struct rtio_sqe *sqe = &ctx->txn_curr->sqe;
+	struct i2c_dt_spec *dt_spec = sqe->iodev->data;
+
+	switch (sqe->op) {
+	case RTIO_OP_RX:
+		return i2c_nrfx_twi_msg_start(dev, I2C_MSG_READ | sqe->iodev_flags,
+					sqe->buf, sqe->buf_len, dt_spec->addr);
+	case RTIO_OP_TINY_TX:
+		return i2c_nrfx_twi_msg_start(dev, I2C_MSG_WRITE | sqe->iodev_flags,
+					sqe->tiny_buf, sqe->tiny_buf_len, dt_spec->addr);
+	case RTIO_OP_TX:
+		return i2c_nrfx_twi_msg_start(dev, I2C_MSG_WRITE | sqe->iodev_flags,
+					sqe->buf, sqe->buf_len, dt_spec->addr);
+	case RTIO_OP_I2C_CONFIGURE:
+		(void)i2c_nrfx_twi_configure(dev, sqe->i2c_config);
+		return false;
+	default:
+		LOG_ERR("Invalid op code %d for submission %p\n", sqe->op, (void *)sqe);
+		return i2c_rtio_complete(ctx, -EINVAL);
+	}
+}
+
+static void i2c_nrfx_twi_complete(const struct device *dev, int status)
+{
+	/** Finalize if there are no more pending xfers */
+	const struct i2c_nrfx_twi_config *config = dev->config;
+	struct i2c_nrfx_twi_data *data = dev->data;
+	struct i2c_rtio *const ctx = data->ctx;
+
+	if (i2c_rtio_complete(ctx, status)) {
+		(void)i2c_nrfx_twi_start(dev);
+	} else {
+		nrfx_twi_disable(&config->twi);
+		k_sem_give(&data->completion_sync);
```

Comment:
i don't think this semaphore is needed anymore, no threads are waiting on it

Reply:
Resolution depends on https://github.com/zephyrproject-rtos/zephyr/pull/72539#discussion_r1596846675